Assertion failure with nested scoped values

On current `master` (77c06727559c3dbb87e5fde46cec87ce59ba61c8) the following example:
```julia
const VALUE = ScopedValue{Int}(1)

x = 0
@with VALUE => 2 begin
    for r = Int[1]
        @with VALUE => 3 begin
            try
                copy(x)
            catch err
                nothing
            end
        end
    end
end
```

triggers an assertion failure
```
Internal error: during type inference of
Tuple{}
Encountered unexpected error in runtime:
MethodError(f=Base.string, args=(Expr(:call, :==, :current_mapping, :val),), world=0x0000000000001720)
jl_method_error_bare at /Users/simon/src/julia/src/gf.c:2215
jl_method_error at /Users/simon/src/julia/src/gf.c:2233
jl_lookup_generic_ at /Users/simon/src/julia/src/gf.c:3145 [inlined]
ijl_apply_generic at /Users/simon/src/julia/src/gf.c:3160
macro expansion at ./error.jl:233 [inlined]
update_scope_mapping! at ./compiler/ssair/passes.jl:1162
sroa_pass! at ./compiler/ssair/passes.jl:1230
run_passes_ipo_safe at ./compiler/optimize.jl:958
run_passes_ipo_safe at ./compiler/optimize.jl:973 [inlined]
optimize at ./compiler/optimize.jl:947
jfptr_optimize_40169 at /Users/simon/src/julia/usr/lib/julia/sys.dylib (unknown line)
_typeinf at ./compiler/typeinfer.jl:262
typeinf at ./compiler/typeinfer.jl:216
typeinf_ext at ./compiler/typeinfer.jl:1125
typeinf_ext_toplevel at ./compiler/typeinfer.jl:1179 [inlined]
typeinf_ext_toplevel at ./compiler/typeinfer.jl:1177
jfptr_typeinf_ext_toplevel_37409 at /Users/simon/src/julia/usr/lib/julia/sys.dylib (unknown line)
jl_apply at /Users/simon/src/julia/src/./julia.h:2174 [inlined]
jl_type_infer at /Users/simon/src/julia/src/gf.c:393
jl_toplevel_eval_flex at /Users/simon/src/julia/src/toplevel.c:932
jl_toplevel_eval_flex at /Users/simon/src/julia/src/toplevel.c:886
ijl_toplevel_eval at /Users/simon/src/julia/src/toplevel.c:952 [inlined]
ijl_toplevel_eval_in at /Users/simon/src/julia/src/toplevel.c:994
eval at ./boot.jl:428 [inlined]
include_string at ./loading.jl:2486
_include at ./loading.jl:2546
include at ./Base.jl:557
jfptr_include_43902 at /Users/simon/src/julia/usr/lib/julia/sys.dylib (unknown line)
exec_options at ./client.jl:318
_start at ./client.jl:526
jfptr__start_67460 at /Users/simon/src/julia/usr/lib/julia/sys.dylib (unknown line)
jl_apply at /Users/simon/src/julia/src/./julia.h:2174 [inlined]
true_main at /Users/simon/src/julia/src/jlapi.c:898
jl_repl_entrypoint at /Users/simon/src/julia/src/jlapi.c:1057
```

Possibly related #52819 

cc: @Keno @vchuravy

Slightly reduced:
```
function f()
	VALUE = ScopedValue(1)

	@with VALUE => 2 begin
		for i = 1
			@with VALUE => 3 begin
				try
					foo()
				catch
					nothing
				end
			end
		end
	end
end
@code_typed f()
ERROR: AssertionError: current_mapping == val
Stacktrace:
  [1] update_scope_mapping!(scope_mapping::Vector{Core.SSAValue}, bb::Int64, val::Core.SSAValue)
    @ Core.Compiler ./compiler/ssair/passes.jl:1162
  [2] sroa_pass!(ir::Core.Compiler.IRCode, inlining::Core.Compiler.InliningState{Core.Compiler.NativeInterpreter})
    @ Core.Compiler ./compiler/ssair/passes.jl:1230
  [3] run_passes_ipo_safe(ci::Core.CodeInfo, sv::Core.Compiler.OptimizationState{…}, caller::Core.Compiler.InferenceResult, optimize_until::Nothing)
    @ Core.Compiler ./compiler/optimize.jl:965
  [4] run_passes_ipo_safe
    @ ./compiler/optimize.jl:980 [inlined]
  [5] optimize(interp::Core.Compiler.NativeInterpreter, opt::Core.Compiler.OptimizationState{…}, caller::Core.Compiler.InferenceResult)
    @ Core.Compiler ./compiler/optimize.jl:954
  [6] _typeinf(interp::Core.Compiler.NativeInterpreter, frame::Core.Compiler.InferenceState)
    @ Core.Compiler ./compiler/typeinfer.jl:262
  [7] typeinf(interp::Core.Compiler.NativeInterpreter, frame::Core.Compiler.InferenceState)
    @ Core.Compiler ./compiler/typeinfer.jl:216
  [8] typeinf_frame(interp::Core.Compiler.NativeInterpreter, mi::Core.MethodInstance, run_optimizer::Bool)
    @ Core.Compiler ./compiler/typeinfer.jl:1018
  [9] typeinf_code(interp::Core.Compiler.NativeInterpreter, mi::Core.MethodInstance, run_optimizer::Bool)
    @ Core.Compiler ./compiler/typeinfer.jl:961
 [10] typeinf_code
    @ ./compiler/typeinfer.jl:955 [inlined]
 [11] code_typed_by_type(tt::Type; optimize::Bool, debuginfo::Symbol, world::UInt64, interp::Core.Compiler.NativeInterpreter)
    @ Base ./reflection.jl:1626
 [12] code_typed_by_type
    @ ./reflection.jl:1606 [inlined]
 [13] code_typed(f::Any, types::Any; kwargs::@Kwargs{})
    @ Base ./reflection.jl:1585
 [14] code_typed(f::Any, types::Any)
    @ Base ./reflection.jl:1580
 [15] macro expansion
    @ ~/Documents/Code/julia/usr/share/julia/stdlib/v1.12/InteractiveUtils/src/macros.jl:230 [inlined]
 [16] top-level scope
    @ REPL[4]:1
Some type information was truncated. Use `show(err)` to see complete types.
```

--- a/base/compiler/ssair/passes.jl
+++ b/base/compiler/ssair/passes.jl
@@ -1225,10 +1225,21 @@ function sroa_pass!(ir::IRCode, inlining::Union{Nothing,InliningState}=nothing)
             bb = compact.active_result_bb - 1
             bbs = scope_mapping[bb]
             if isexpr(stmt, :leave) && bbs != SSAValue(0)
-                update_scope_mapping!(scope_mapping, bb+1, scope_mapping[block_for_inst(compact, bbs)])
-            else
-                update_scope_mapping!(scope_mapping, bb+1, bbs)
+                # Here we want to count the number of scopes that we're leaving,
+                # which is the same as the number of EnterNodes being referenced
+                # by `stmt.args`. Which have :scope set. In practice, the frontend
+                # does emit these in order, so we could simply go to the last one,
+                # but we want to avoid making that semantic assumption.
+                for i = 1:length(stmt.args)
+                    scope = stmt.args[i]
+                    scope === nothing && continue
+                    enter = compact[scope][:inst]
+                    @assert isa(enter, EnterNode)
+                    isdefined(enter, :scope) || continue
+                    bbs = scope_mapping[block_for_inst(compact, bbs)]
+                end
             end
+            update_scope_mapping!(scope_mapping, bb+1, bbs)
         end
         # check whether this statement is `getfield` / `setfield!` (or other "interesting" statement)
         is_setfield = is_isdefined = is_finalizer = is_keyvalue_get = false

--- a/test/compiler/irpasses.jl
+++ b/test/compiler/irpasses.jl
@@ -1801,3 +1801,22 @@ let n = 1000
     end
 end
 @test f_1000_blocks() == 0
+
+# https://github.com/JuliaLang/julia/issues/53521
+# Incorrect scope counting in :leave
+using Base.ScopedValues
+function f53521()
+    VALUE = ScopedValue(1)
+    @with VALUE => 2 begin
+        for i = 1
+            @with VALUE => 3 begin
+                try
+                    foo()
+                catch
+                    nothing
+                end
+            end
+        end
+    end
+end
+@test code_typed(f53521)[1][2] === Nothing

